### PR TITLE
Add option to detect duplicate values

### PR DESCRIPTION
**Pull request by hugovk**
Implements https://github.com/legoktm/jsonchecker/issues/5.
- Ignore *.pyc files generated when running tests. Updated with common things from https://github.com/github/gitignore
- Add a more complex JSON in bad3.json to test duplicate keys a bit more thoroughly.
- Add DuplicateValueFinder to find duplicate values.
  - Add tests for this with test data.
    - Also add non-duplicate values in good.json.
  - Refactor DuplicateKeyFinder and DuplicateValueFinder so they have a common base class DuplicateFinder to remove duplicate code.
  - Similarly for tests, refactor DuplicateKeyFinderTest and DuplicateValueFinderTest so they have a common base class DuplicateFinderTest. DuplicateKeyFinder to remove duplicate code.
    - Also add expected_error_helper to reduce duplicate test code.

Usage for checking duplicate keys is unchanged.

To also check for duplicate values as well as keys, use the `--values` option. Then if either of these checks fail, the exit code is 1.

### CODE REVIEW

**Comment by jayvdb** [comment]
Could you submit the first three commits as one separate PR.  That can be merged quickly.

**Comment by hugovk** [comment]
@jayvdb Done: PR https://github.com/legoktm/jsonchecker/pull/10

**Comment by jayvdb** [comment]
Would you please rebase and squash these commits as one commit on master?

**Comment by hugovk** [comment]
Done.

File: jsonchecker/__init__.py [hunk 1]
```diff
@@ -91,6 +81,71 @@ def exit(self):
...
+        """Determine whether `v` can be hashed."""
+        try:
+            hash(v)
+        except TypeError:
+            return False
+        return True
+
+    def dupes_in_list(self, l):
+        """Return hashable duplicates from this list."""
+        seen = set()
```
Comment (jayvdb) [inline jsonchecker/__init__.py]: you could use `collections.Counter` for the seen lists, especially if it useful to know how many times a value appears, such as ordering them errors by number of occurrences.

Sorry I forgot to mention this earlier.  But it isnt a significant issue if you dont have time to attempt it.  I'll merge this tomorrow either way.

**Comment by hugovk** [comment]
I did some timings with:
```python

def hashable(v):
    """Determine whether `v` can be hashed."""
    try:
        hash(v)
    except TypeError:
        return False
    return True

def test_dupes_in_list_with_sets(l):
    """Return hashable duplicates from this list."""
    seen = set()
    seen_twice = set()
    # Adds all elements it doesn't know yet to seen and
    # adds all others to seen_twice
    for x in l:
        if hashable(x):
            if x in seen:
                seen_twice.add(x)
            else:
                seen.add(x)
    return list(seen_twice)

from collections import Counter

def test_dupes_in_list_with_counter(l):
    """Return hashable duplicates from this list."""
    stuff = [t for t in l if hashable(t)]
    return [item for item, count in Counter(stuff).items() if count > 1]

import sys, timeit
print (sys.version + "\n")
xs = list(range(10000)) + [999]
fns = [p for name, p in globals().items() if name.startswith('test')]
for fn in fns:
    assert fn(xs) == [999]
    print('%50s %.5f' % (fn, timeit.timeit(lambda: fn(xs), number=1000)))

```

And got:

```
2.7.10 (default, Sep 23 2015, 04:34:21) 
[GCC 4.2.1 Compatible Apple LLVM 7.0.0 (clang-700.0.72)]

<function test_dupes_in_list_with_sets at 0x10d9012a8> 4.35807
<function test_dupes_in_list_with_counter at 0x10d91e5f0> 7.35557
```

So I guess sets win for the current functionality.

---

Would you like tests?

```python
+    def test_dupes_in_list_found(self):
+        """Test returns duplicates from a list."""
+        input = ["a", "b", "a"]
+        output = self.finder.dupes_in_list(input)
+        self.assertEqual(output, ["a"])
+
+    def test_dupes_in_list_not_found(self):
+        """Test returns empty list when no duplicates."""
+        input = ["a", "b", "c"]
+        output = self.finder.dupes_in_list(input)
+        self.assertEqual(output, [])
```

File: README.rst [hunk 2]
```diff
@@ -37,4 +37,8 @@ of the full filename for each file scanned. Useful if you plan on checking a lar
 The script will exit with a status code of 1 if any duplicate keys are found,
 0 if none are.
 
+An optional ``--values`` argument can be passed to additionally check for duplicate values within each key.
```
Comment (jayvdb) [inline README.rst]: trailing whitespace on these two lines

**Comment by jayvdb** [comment]
Thx for doing timing with Counter.  No extra tests needed.

**Comment by jayvdb** [comment]
Committed with RST whitespace fixed as ea605f9

**Comment by jayvdb** [comment]
@hugovk, did you look at adding this feature to https://github.com/dmeranda/demjson ?
Did you find problems with `demjson` that pushed you towards using `jsonchecker` instead?

**Comment by hugovk** [comment]
If I remember correctly, demjson only checks JSON that's strictly correct, and strictly speaking duplicate things are allowed in json. I think I found jsonchecker via the python-list post on the README (which by the way now leads to a 404) or some Stack Overflow pointing to it or this project.

But now I look again I see demjson does have options to warn or forbid duplicates keys!

**Pull request closed without merging**